Домашнее задание

xafish opened the pull request:
hw02
Конфигурирование Spring-приложений

xafish committed:
Домашнее задание
hw02
Конфигурирование Spring-приложений

kataus commented on hw02/src/main/java/ru/otus/spring/config/AppConfig.java:
```diff
@@ -0,0 +1,15 @@
+
+@Configuration
+public class AppConfig {
+    @Bean
```
Вручную создаем бины только в крайнем случае, если требуется какое сложная необычная настройка. Но всех остальных случаях предпочтение отдается аннотациям над классом

kataus commented on hw02/src/main/java/ru/otus/spring/dao/QuestionsDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Getter
```
А зачем он нам тут? Появится публичный метод, не заявленный в интерфейсе.

kataus commented on hw02/src/main/java/ru/otus/spring/dao/QuestionsDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+    public QuestionsDaoImpl(FileManager fileManager) {
+        questions = new HashMap<>();
+
+        List<String[]> csvArray = fileManager.getCsvArray();
```
Тяжелому функционалу загрузки не место в конструкторе. Spring и так заслуженно критикуют за долгий старт, а тут ещё утяжеляется. Загрузка по необходимости, можно с кэшированием.

kataus commented on hw02/src/main/java/ru/otus/spring/dao/QuestionsDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public void runTest(){
```
DAO - Data Access Object. Бизнес-логике тут не место

kataus commented on hw02/src/main/java/ru/otus/spring/dao/QuestionsDaoImpl.java:
```diff
@@ -0,0 +1,64 @@
+                }
+        );
+        in.close();
+        System.out.printf((sNum.get()==questions.size()?"Well done ":"Good job ")+username+"!\n"
```
Давайте сделаем тест на данный метод. Сейчас это будет сложно, что является признаком проблем в его архитектуре

kataus commented on hw02/src/main/java/ru/otus/spring/utils/FileManager.java:
```diff
@@ -0,0 +1,40 @@
+
+@AllArgsConstructor
+@Service
+public class FileManager {
```
Интерфейсы - очень сильный инструмент первичной абстракции кода. Их активное использование очень удобно и реально помогает строить гибкий код. Не стоит ими пренебрегать

kataus commented on hw02/src/test/java/QuestionDaoTest.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Test
+    public void test(){
+        Assertions.assertTrue(questionsDao.findByQuestions("question1").contains("q1_text1"));
```
Давайте начнем с unit-тестов, интеграционные стоит писать в меньшем объеме, после того как unit-тестами покрыто

xafish committed:
Домашнее задание
hw02
Конфигурирование Spring-приложений
Внёс правки по комментариям.

xafish committed:
Домашнее задание
hw02
Конфигурирование Spring-приложений
Внёс правки по комментариям.
+ добавил тестирование вывода

kataus commented on hw02/src/main/java/ru/otus/spring/service/TestServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+        PrintStream out = ioServiceStreams.out();
+        InputStream in = ioServiceStreams.in();
+
+        Scanner inScanner = new Scanner(in);
```
Будет сложно тестировать, работу с консолью тоже лучше убрать в ioService
